crates/bevy_ecs/src/system/system_registry.rs

```diff
@@ -0,0 +1,324 @@
+use std::any::TypeId;
+use std::marker::PhantomData;
+
+use crate::system::{BoxedSystem, Command, IntoSystem};
+use crate::world::{Mut, World};
+// Needed for derive(Component) macro
+use crate::{self as bevy_ecs, TypeIdMap};
+use bevy_ecs_macros::Resource;
+
+/// Stores initialized [`System`]s, so they can be reused and run in an ad-hoc fashion.
+///
+/// Systems are keyed by their [`SystemId`]:
+///  - repeated calls with the same function type will reuse cached state, including for change detection
+///
+/// Any [`Commands`](crate::system::Commands) generated by these systems (but not other systems), will immediately be applied.
+///
+/// This type is stored as a [`Resource`](crate::system::Resource) on each [`World`], initialized by default.
+/// However, it will likely be easier to use the corresponding methods on [`World`],
+/// to avoid having to worry about split mutable borrows yourself.
+///
+/// # Limitations
+///
+///  - stored systems cannot be chained: they can neither have an [`In`](crate::system::In) nor return any values
+///  - stored systems cannot recurse: they cannot run other systems via the [`SystemRegistry`] methods on `World` or `Commands`
+///  - exclusive systems cannot be used
+///
+/// # Examples
+///
+/// You can run a single system directly on the World,
+/// applying its effect and caching its state for the next time
+/// you call this method.
+///
+/// ```rust
+/// use bevy_ecs::prelude::*;
+///
+/// let mut world = World::new();  
+///
+/// #[derive(Default, PartialEq, Debug)]
+/// struct Counter(u8);
+///
+/// fn count_up(mut counter: ResMut<Counter>){
+///     counter.0 += 1;
+/// }
+///
+/// world.init_resource::<Counter>();
+/// world.run_system(count_up);
+///
+/// assert_eq!(Counter(1), *world.resource());
+/// ```
+///
+/// These systems immediately apply commands and cache state,
+/// ensuring that change detection and [`Local`](crate::system::Local) variables work correctly.
+///
+/// ```rust
+/// use bevy_ecs::prelude::*;
+///
+/// let mut world = World::new();
+///
+/// #[derive(Component)]
+/// struct Marker;
+///
+/// fn spawn_7_entities(mut commands: Commands) {
+///     for _ in 0..7 {
+///         commands.spawn(Marker);
+///     }
+/// }
+///
+/// fn assert_7_spawned(query: Query<(), Added<Marker>>){
+///     let n_spawned = query.iter().count();
+///     assert_eq!(n_spawned, 7);
+/// }
+///
+/// world.run_system(spawn_7_entities);
+/// world.run_system(assert_7_spawned);
+/// ```
+#[derive(Resource, Default)]
+pub struct SystemRegistry {
+    systems: TypeIdMap<(bool, BoxedSystem)>,
+}
+
+impl SystemRegistry {
+    /// Registers a system in the [`SystemRegistry`], so it can be run later.
+    ///
+    /// Repeatedly registering a system will have no effect.
```

Comment:
This will be a problem for closures that capture local data. Here's a strawman example that I just ran and tested:
```
use bevy::prelude::*;

fn main() {
    App::new()
        .add_plugins(DefaultPlugins)
        .insert_resource(Count(0))
        .add_systems(Update, shooter)
        .run();
}

#[derive(Resource)]
struct Count(u32);

fn shooter(
    mut commands: Commands,
    mut count: ResMut<Count>,
) {
    count.0 += 1;
    if count.0 % 100 == 0 {
        let c = count.0;
        commands.run_system(move || println!("{c}"));
    }
}
```
The output of this is
```
100
100
100
100
...etc...
```

I'd like to use one-shot systems to implement a framework for reactive programming patterns inside of Bevy's ECS that can use closures, but it seems that this approach to a system registry will not be suitable. If there were an API that allowed associating each run request with an ID or entity as suggested by @maniwani in [this thread](https://github.com/bevyengine/bevy/pull/8963/files#r1271546245) that would probably help. Otherwise I expect I will need to reimplement one-shot systems for the framework to be feasible.

Reply:
This example definitely made me think closure TypeId behavior is not predictable enough to use as a basis for an API.

Reply:
I've updated the code to solve this. Now every time run_system is called, the system is added, run, and removed from the registry, ensuring that local data is recaptured each time.

(I haven't tested your example in particular because it's 1:40 AM)